libzmix/src/signatures/delg_cred_cdd/groth_sig.rs

```diff
@@ -0,0 +1,439 @@
+            let S = &self.S * &r_prime_inv;
+            Self {
+                R,
+                S,
+                T: self.T.scaled_by(&r_prime_inv),
+            }
+        }
+    }
+}
+
+impl_GrothS!(
+    Groth1SetupParams,
+    Groth1Verkey,
+    Groth1Sig,
+    GrothS1,
+    G2,
+    G1,
+    G1Vector
+);
+
+impl_GrothS!(
+    Groth2SetupParams,
+    Groth2Verkey,
+    Groth2Sig,
+    GrothS2,
+    G1,
+    G2,
+    G2Vector
+);
+
+/// Returns tuple of groups elements where the elements are result of scalar multiplication involving the same field element. Uses w-NAF
+#[macro_export]
+macro_rules! var_time_mul_scl_mul_with_same_field_element {
+    ( $group:ident, $field_elem:expr, $( $group_elem_table:expr ),* ) => {{
+        let wnaf = $field_elem.to_wnaf(5);
+        (
+            $(
+            $group::wnaf_mul(&$group_elem_table, &wnaf),
+            )*
+        )
+    }}
+}
+
+impl GrothS1 {
+    impl_GrothS_setup!(Groth1SetupParams, G1, G1Vector);
+
+    pub fn keygen(setup_params: &Groth1SetupParams) -> (GrothSigkey, Groth1Verkey) {
+        let sk = FieldElement::random();
+        let vk = &setup_params.g2 * &sk;
+        (GrothSigkey(sk), Groth1Verkey(vk))
+    }
+}
+
+impl Groth1Sig {
+    pub fn new(
+        messages: &[G1],
+        sk: &GrothSigkey,
+        setup_params: &Groth1SetupParams,
+    ) -> DelgResult<Self> {
+        impl_GrothSig_new!(
+            messages,
+            sk,
+            setup_params.y,
+            setup_params.g2,
+            setup_params.g1,
+            G1Vector
+        )
+    }
+
+    impl_GrothSig_randomize!();
+
+    pub fn verify(
+        &self,
+        messages: &[G1],
+        verkey: &Groth1Verkey,
+        setup_params: &Groth1SetupParams,
+    ) -> DelgResult<bool> {
+        if messages.len() > setup_params.y.len() {
+            return Err(DelgError::UnsupportedNoOfMessages {
+                expected: setup_params.y.len(),
+                given: messages.len(),
+            });
+        }
+
+        // e(S, R) == e(y_0, g2) * (g1, V) => e(y_0, g2) * (g1, V) * e(S, R)^-1 == 1 => e(y_0, g2) * (g1, V) * e(S^-1, R) == 1
+        let negS = self.S.negation();
+        let e0 = GT::ate_multi_pairing(vec![
+            (&setup_params.y[0], &setup_params.g2),
+            (&setup_params.g1, &verkey.0),
+            (&negS, &self.R),
+        ]);
+        if !e0.is_one() {
+            return Ok(false);
+        }
+
+        let negR = self.R.negation();
+        for i in 0..messages.len() {
+            // e(T_i, R) == e(m_i, g2) * e(y_i, V) => 1 == e(m_i, g2) * e(y_i, V) * e(T_i, R)^-1 = 1 == e(m_i, g2) * e(y_i, V) * e(T_i, R^-1)
+            let e = GT::ate_multi_pairing(vec![
+                (&messages[i], &setup_params.g2),
+                (&setup_params.y[i], &verkey.0),
+                (&self.T[i], &negR),
+            ]);
+            if !e.is_one() {
+                return Ok(false);
+            }
+        }
+        Ok(true)
+    }
+
+    pub fn verify_fast(
```

Comment:
Besides the math difference for speed, it's not obvious when I should use `verify` vs `verify_fast`. Is `verify` in constant time or just a naive implementation for verification?

If `verify_fast` is also constant time, could we just have a single verify function.

Reply:
`verify_fast` is an argument of knowledge. Its explained in the RFC